sklearn/tests/test_common_non_default.py

```diff
@@ -0,0 +1,254 @@
+import pprint
+import itertools
+from inspect import signature
+from typing import Optional, List, Dict, Any
+
+from sklearn.datasets import load_iris
+from sklearn.base import is_classifier, is_regressor
+from sklearn.tree._classes import BaseDecisionTree
+from sklearn.ensemble._forest import BaseForest
+from sklearn.ensemble._gb import BaseGradientBoosting
+from sklearn.utils import all_estimators
+from sklearn.utils.estimator_checks import (
+    _enforce_estimator_tags_y,
+    parametrize_with_checks,
+)
+
+
+categorical_params = [
+    "solver",
+    "algorithm",
+    "loss",
+    "strategy",
+    "selection",
+    "criterion",
+    "multi_class",
+    "kernel",
+]
+bool_params = [
+    "fit_prior",
+    "fit_intercept",
+    "positive",
+    "normalize",
+    "dual",
+    "average",
+    "shuffle",
+]
```

Comment:
how about using type hints and getting these from there?

Reply:
Yeah, we could probably try to auto-detect this in the future, but this PR has already a lot of moving parts, and I prefer to manually list parameters to check for now.

Reply:
There are two aspects to my comment which I kinda failed to better explain:

- it's nice to get the info from type hints cause it's nice ;)
- it allows better extending this functionality to third party estimators.

My personal motivation behind my comment was the second one. I'm happy to leave these hard coded here instead of adding type hints in our estimators in this PR, but it would be really nice if this PR also introduces detecting the type from the type hints, unless you think it'd make this PR too big/complicated to review.

Reply:
I have looked into auto-detecting it at least for boolean, but then you endup with a long list of parameters to exclude, e.g.
```
['verbose', 'copy', 'copy_X', 'warm_start', 'compute_score', 'compute_labels', 'validate', 'accept_sparse', 'check_inverse', 'copy_X_train', 'add_indicator', 'skip_complete', 'fit_path']
```
and I have only done estimators starting between A to D. Each of those shouldn't impact the model results but would add to the total runtime. So I think it's easier to whitelist parameter names rather than blacklist them in this case. It could be still possible to use this in contrib projects, by passing the whitelist as input parameter to `detect_all_params` function (or just monkeypatching the `categorical_params` / `bool_params` lists for now).